scalding-core/src/main/scala/com/twitter/scalding/serialization/Serialization.scala

```diff
@@ -0,0 +1,149 @@
+/*
+Copyright 2015 Twitter, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package com.twitter.scalding.serialization
+
+import java.io.{ ByteArrayInputStream, ByteArrayOutputStream, InputStream, OutputStream, Serializable }
+
+import scala.util.{ Failure, Success, Try }
+import scala.util.hashing.Hashing
+
+/**
+ * This is a base Input/OutputStream-based serialization typeclass
+ * This is useful for value serialization in hadoop when we don't
+ * need to do key sorting for partitioning.
+ *
+ * This serialization typeclass must serialize equivalent objects
+ * identically to be lawful. Given that constraint, we can always
+ * get an Equiv and Hashing from a Serialization (by doing byte-wise
+ * equivalence or byte-wise hashing).
+ *
+ * A serialization always gives a hash because one can just
+ * serialize and then hash the bytes. You might prefer another
+ * implementation. This must satisfy:
+ *   (!equiv(a, b)) || (hash(a) == hash(b))
+ */
+trait Serialization[T] extends Equiv[T] with Hashing[T] with Serializable {
+  def read(in: InputStream): Try[T]
+  def write(out: OutputStream, t: T): Try[Unit]
+  /**
+   * If all items have a static size, this returns Some, else None
+   */
+  def staticSize: Option[Int] = None
+  /**
+   * returns Some if the size is cheap to calculate.
+   * otherwise the caller should just serialize into an ByteArrayOutputStream
+   */
+  def dynamicSize(t: T): Option[Int] = None
+}
+
+object Serialization {
+  import JavaStreamEnrichments._
+  /**
+   * This is a constant for us to reuse in Serialization.write
+   */
+  val successUnit: Try[Unit] = Success(())
+
+  def equiv[T](a: T, b: T)(implicit ser: Serialization[T]): Boolean =
+    ser.equiv(a, b)
+
+  def hash[T](t: T)(implicit ser: Serialization[T]): Int =
+    ser.hash(t)
+
+  def read[T](in: InputStream)(implicit ser: Serialization[T]): Try[T] =
+    ser.read(in)
+
+  def write[T](out: OutputStream, t: T)(implicit ser: Serialization[T]): Try[Unit] =
+    ser.write(out, t)
+
+  def toBytes[T: Serialization](t: T): Array[Byte] = {
```

Comment:
an instance of Serialization could do this better if it knew the size to allocate ByteArrayOutputStream right?